fem/nonlinearform_ext.hpp

```diff
@@ -17,28 +17,75 @@
 
 namespace mfem
 {
+
 class NonlinearForm;
+class NonlinearFormIntegrator;
 
+/** @brief Class extending the NonlinearForm class to support the different
+    AssemblyLevel%s. */
 class NonlinearFormExtension : public Operator
 {
 protected:
-   NonlinearForm *n; ///< Not owned
+   const NonlinearForm *nlf; ///< Not owned
+
 public:
-   NonlinearFormExtension(NonlinearForm *form);
-   virtual void AssemblePA() = 0;
+   NonlinearFormExtension(const NonlinearForm*);
+
+   /// Assemble at the AssemblyLevel of the subclass.
+   virtual void Assemble() = 0;
+   /// Assemble gradient data at the AssemblyLevel of the subclass, for the
+   /// state @a x, which is assumed to be a ldof Vector.
+   virtual void AssembleGradient(const Vector &x) = 0;
+
+   /// Assumes that @a x is a ldof Vector.
+   virtual Operator &GetGradient(const Vector &x) const = 0;
```

Comment:
Do we need both of these methods? `AssembleGradient` seems redundant because `GetGradient` also takes `x` as input.

Reply:
This is related to having separate `NonlinearForm::SetupGradient` and `NonlinearForm::GetGradient` for the PA case. I thought we decided to have the option to do quadrature computations separately.